parlai/core/worlds.py

```diff
@@ -1499,6 +1531,629 @@ def shutdown(self):
+            idx:       the worker index that produced the batch
+        """
+        batch_obs: Union[List[Message], Batch] = None
+        queue_signal: QueueSignal = None
+        worker_idx: int = None
+        while queue_signal != QueueSignal.BATCH and not self.epoch_done():
+            queue_signal, batch_obs, worker_idx = self.produce_queue.get()
+            if queue_signal == QueueSignal.WORKER_FINISHED:
+                self.finished_workers.append(worker_idx)
+        if queue_signal == QueueSignal.BATCH:
+            self.num_consume += 1
+        return queue_signal, batch_obs, worker_idx
+
+    def _maybe_handle_buffers(self, batch_obs: Batch, worker_idx: int) -> Batch:
+        """
+        Handle batch_obs buffers.
+
+        Essentially, housekeeping on the buffers used for the given `worker_idx`.
+            1. Update QueueWorld buffers for `worker_idx`
+            2. Resize QueueWorld buffers for `worker_idx`
+            3. Set appropriate batch attrs with QueueWorld buffers
+
+        :param batch_obs:
+            The batch of observations
+
+        :param worker_idx:
+            The worker from which the batch came
+
+        :return: batch_obs, a batch post-buffer-housekeeping
+        :rtype: Batch
+        """
+        if hasattr(self.world.get_model_agent(), 'batchify'):
+            for key, buff in batch_obs.updated_buffers.items():
+                # Update
+                self.buffers[worker_idx][key] = buff
+            for key, new_size in batch_obs.resized.items():
+                self.buffers[worker_idx][key].resize_(new_size)
+            for key, buff in self.buffers[worker_idx].items():
+                setattr(batch_obs, key, buff)
+        return batch_obs
+
+    def parley(self):
+        """
+        QueueWorld parley.
+
+        QueueWorld receives a batch from its produce queue, gives it to an agent to act,
+        then puts it back on the appropriate consume queue.
+        """
+        self._maybe_begin_processes()
+        self._maybe_parley_init()
+        queue_signal, batch_obs, worker_idx = self._poll_produce_queue()
+        if self.epoch_done():
+            return
+        batch_obs = self._maybe_handle_buffers(batch_obs, worker_idx)
+        batch_acts = self.batch_act(batch_obs)
+        self.consume_queues[worker_idx].put_nowait((QueueSignal.BATCH, batch_acts))
+        self.update_counters(batch_obs)
+
+    def batch_act(
+        self, batch_observation: Union[List[Message], Batch]
+    ) -> List[Message]:
+        """
+        Batch act on agent.
+
+        :param batch_observation:
+            The observation that the agent will act upon
+
+        :return: batch_actions
+        :rtype: List[Message]
+        """
+        a = self.world.get_model_agent()
+        if hasattr(a, 'batch_act'):
+            batch_actions = a.batch_act(batch_observation)
+        else:
+            batch_actions = []
+            for w in self.worlds:
+                agents = w.get_agents()
```

Comment:
i'm a bit confused by this loop here. what is agent[1]?

Reply:
ahh good catch, this should just be the `world.get_model_agent()`. 

To be honest, I have not even considered supporting agents who do not have batch act, and i'm not sure it makes sense to. So this branch of code is thus far not even touched